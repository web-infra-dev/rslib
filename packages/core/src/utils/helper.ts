import fs from 'node:fs';
import fsP from 'node:fs/promises';
import path from 'node:path';
import color from 'picocolors';
import type { PkgJson } from '../types';
import { logger } from './logger';

/**
 * Node.js built-in modules.
 * Copied from https://github.com/webpack/webpack/blob/dd44b206a9c50f4b4cb4d134e1a0bd0387b159a3/lib/node/NodeTargetPlugin.js#L12-L72
 */
export const nodeBuiltInModules: Array<string | RegExp> = [
  'assert',
  'assert/strict',
  'async_hooks',
  'buffer',
  'child_process',
  'cluster',
  'console',
  'constants',
  'crypto',
  'dgram',
  'diagnostics_channel',
  'dns',
  'dns/promises',
  'domain',
  'events',
  'fs',
  'fs/promises',
  'http',
  'http2',
  'https',
  'inspector',
  'inspector/promises',
  'module',
  'net',
  'os',
  'path',
  'path/posix',
  'path/win32',
  'perf_hooks',
  'process',
  'punycode',
  'querystring',
  'readline',
  'readline/promises',
  'repl',
  'stream',
  'stream/consumers',
  'stream/promises',
  'stream/web',
  'string_decoder',
  'sys',
  'timers',
  'timers/promises',
  'tls',
  'trace_events',
  'tty',
  'url',
  'util',
  'util/types',
  'v8',
  'vm',
  'wasi',
  'worker_threads',
  'zlib',
  /^node:/,

  // cspell:word pnpapi
  // Yarn PnP adds pnpapi as "builtin"
  'pnpapi',
];

export async function calcLongestCommonPath(
  absPaths: string[],
): Promise<string | null> {
  if (absPaths.length === 0) {
    return null;
  }

  // we support two cases
  // 1. /packages-a/src/index.ts
  // 2. D:/packages-a/src/index.ts
  const sep = path.posix.sep as '/';

  const splitPaths = absPaths.map((p) => p.split(sep));
  let lcaFragments = splitPaths[0]!;
  for (let i = 1; i < splitPaths.length; i++) {
    const currentPath = splitPaths[i]!;
    const minLength = Math.min(lcaFragments.length, currentPath.length);

    let j = 0;
    while (j < minLength && lcaFragments[j] === currentPath[j]) {
      j++;
    }

    lcaFragments = lcaFragments.slice(0, j);
  }

  let lca = lcaFragments.length > 0 ? lcaFragments.join(sep) : sep;

  const stats = await fsP.stat(lca);
  if (stats?.isFile()) {
    lca = path.dirname(lca);
  }

  return lca;
}

export const readPackageJson = (rootPath: string): undefined | PkgJson => {
  const pkgJsonPath = path.join(rootPath, './package.json');

  if (!fs.existsSync(pkgJsonPath)) {
    logger.warn(`package.json does not exist in the ${rootPath} directory`);
    return;
  }

  try {
    return JSON.parse(fs.readFileSync(pkgJsonPath, 'utf8'));
  } catch (err) {
    logger.warn(`Failed to parse ${pkgJsonPath}, it might not be valid JSON`);
    return;
  }
};

export const isObject = (obj: unknown): obj is Record<string, any> =>
  Object.prototype.toString.call(obj) === '[object Object]';

export const isEmptyObject = (obj: object): boolean => {
  return Object.keys(obj).length === 0;
};

export function pick<T, U extends keyof T>(
  obj: T,
  keys: ReadonlyArray<U>,
): Pick<T, U> {
  return keys.reduce(
    (ret, key) => {
      if (obj[key] !== undefined) {
        ret[key] = obj[key];
      }
      return ret;
    },
    {} as Pick<T, U>,
  );
}

export function omit<T extends object, U extends keyof T>(
  obj: T,
  keys: ReadonlyArray<U>,
): Omit<T, U> {
  return Object.keys(obj).reduce(
    (ret, key) => {
      if (!keys.includes(key as U)) {
        ret[key as keyof Omit<T, U>] = obj[key as keyof Omit<T, U>];
      }
      return ret;
    },
    {} as Omit<T, U>,
  );
}

export { color };
