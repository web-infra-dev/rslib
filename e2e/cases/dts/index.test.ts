import { join } from 'node:path';
import { buildAndGetResults } from '@e2e/helper';
import { describe, expect, test } from 'vitest';

describe('dts when bundle: false', () => {
  test('basic', async () => {
    const fixturePath = join(__dirname, 'bundle-false');
    const { files, contents } = await buildAndGetResults(
      fixturePath,
      'rslib.config.ts',
      'dts',
    );

    expect(files.esm).toMatchInlineSnapshot(`
      [
        "./dist/esm/index.d.ts",
        "./dist/esm/sum.d.ts",
        "./dist/esm/utils/numbers.d.ts",
        "./dist/esm/utils/strings.d.ts",
      ]
    `);
    expect(contents.esm).toMatchSnapshot();
  });

  test('dts false', async () => {
    const fixturePath = join(__dirname, 'bundle-false');
    const { files } = await buildAndGetResults(
      fixturePath,
      'dtsFalse.config.ts',
      'dts',
    );

    expect(files.esm).toBe(undefined);
  });

  test('distPath', async () => {
    const fixturePath = join(__dirname, 'bundle-false');
    const { files } = await buildAndGetResults(
      fixturePath,
      'distPath.config.ts',
      'dts',
    );

    expect(files.esm).toMatchInlineSnapshot(`
      [
        "./dist/custom/index.d.ts",
        "./dist/custom/sum.d.ts",
        "./dist/custom/utils/numbers.d.ts",
        "./dist/custom/utils/strings.d.ts",
      ]
    `);
  });

  test('abortOnError: false', async () => {
    const fixturePath = join(__dirname, 'bundle-false');
    const { isSuccess } = await buildAndGetResults(
      fixturePath,
      'abortOnError.config.ts',
      'dts',
    );

    expect(isSuccess).toBe(true);
  });

  test('autoExtension: true', async () => {
    const fixturePath = join(__dirname, 'bundle-false');
    const { files } = await buildAndGetResults(
      fixturePath,
      'autoExtension.config.ts',
      'dts',
    );

    expect(files.cjs).toMatchInlineSnapshot(`
      [
        "./dist/cjs/index.d.cts",
        "./dist/cjs/sum.d.cts",
        "./dist/cjs/utils/numbers.d.cts",
        "./dist/cjs/utils/strings.d.cts",
      ]
    `);
  });
});

describe('dts when bundle: true', () => {
  test('basic', async () => {
    const fixturePath = join(__dirname, 'bundle');
    const { entryFiles, entries } = await buildAndGetResults(
      fixturePath,
      'rslib.config.ts',
      'dts',
    );

    expect(entryFiles.esm).toEqual('./dist/esm/index.d.ts');
    expect(entries).toMatchSnapshot();
  });

  test('dts false', async () => {
    const fixturePath = join(__dirname, 'bundle');
    const { entryFiles } = await buildAndGetResults(
      fixturePath,
      'dtsFalse.config.ts',
      'dts',
    );

    expect(entryFiles.esm).toEqual(undefined);
  });

  test('distPath', async () => {
    const fixturePath = join(__dirname, 'bundle');
    const { entryFiles } = await buildAndGetResults(
      fixturePath,
      'distPath.config.ts',
      'dts',
    );

    expect(entryFiles.esm).toEqual('./dist/custom/index.d.ts');
  });

  test('abortOnError: false', async () => {
    const fixturePath = join(__dirname, 'bundle');
    const { isSuccess } = await buildAndGetResults(
      fixturePath,
      'abortOnError.config.ts',
      'dts',
    );

    expect(isSuccess).toBe(true);
  });

  test('autoExtension: true', async () => {
    const fixturePath = join(__dirname, 'bundle');
    const { entryFiles } = await buildAndGetResults(
      fixturePath,
      'autoExtension.config.ts',
      'dts',
    );

    expect(entryFiles.cjs).toEqual('./dist/cjs/index.d.cts');
  });
});
