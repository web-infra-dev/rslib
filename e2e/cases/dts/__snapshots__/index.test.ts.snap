// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`dts when bundle: false > basic 1`] = `
{
  "./dist/esm/index.d.ts": "export * from './utils/numbers';
export * from './utils/strings';
export * from './sum';
",
  "./dist/esm/sum.d.ts": "export declare const numSum: number;
export declare const strSum: string;
",
  "./dist/esm/utils/numbers.d.ts": "export declare const num1 = 1;
export declare const num2 = 2;
export declare const num3 = 3;
",
  "./dist/esm/utils/strings.d.ts": "export declare const str1 = "str1";
export declare const str2 = "str2";
export declare const str3 = "str3";
",
}
`;

exports[`dts when bundle: true > basic 1`] = `
{
  "esm": "export declare const num1 = 1;

export declare const num2 = 2;

export declare const num3 = 3;

export declare const numSum: number;

export declare const str1 = "str1";

export declare const str2 = "str2";

export declare const str3 = "str3";

export declare const strSum: string;

export { }
",
}
`;
